Limit how many threads can prepare a statement

tgeoghegan opened the pull request:
`StatementCache` aims to prevent redundant SQL statement preparation with a simple cache. However in cases where many, many writers are racing to prepare a statement, lots of redundant and expensive work can be done because `tokio_postgres::client::Client::prepare_typed` is called before `StatementCache::insert`, and only the latter takes the lock on the cache.

The obvious thing to do here would be to call `prepare_typed` while holding the write lock on the cache. We can't do this because `prepare_typed` is async, and the `std::sync::RwLock` used by `StatementCache` can't be held across the `await`. Changing the lock to `tokio::sync::RwLock` won't work because it forces a large part of `deadpool-postgres` to become `async` and in particular makes it impossible for certain types to implement `Drop`.

This commit allows a `StatementCache` value to be either a `Statement` or an `Arc<tokio::sync::Semaphore>`. This way, the `RwLock` can remain synchronous and then writers can acquire and hold the `Semaphore` across the `await`.

MuhtasimTanmoy committed:
docs(redis): add sentinel example and config usage to README

Include a Rust example demonstrating how to use Redis Sentinel with the `deadpool_redis` crate. Also, add an example showing how to configure Redis Sentinel using the `config` and `dotenvy` crates.

tgeoghegan committed:
Limit how many threads can prepare a statement

`StatementCache` aims to prevent redundant SQL statement preparation
with a simple cache. However in cases where many, many writers are
racing to prepare a statement, lots of redundant and expensive work can
be done because `tokio_postgres::client::Client::prepare_typed` is
called before `StatementCache::insert`, and only the latter takes the
lock on the cache.

This commit allows entries in the `StatementCache` map to be either
empty, a Semaphore, or a prepared Statement. The first call to
`prepare_typed()` will insert a `Semaphore`. Then, tasks will wait for a
semaphore permit before preparing the query, to ensure that only one
task sends a `PREPARE` statement at once. At both steps in the process,
the map will be read via a read lock first, then again via a write lock,
before updating the entry.

divergentdave commented:
Here's a test I wrote up to check that this behaves properly in the face of cancellation. The result looks good, cancelling one future drops its semaphore permit, then the other future gets a permit and proceeds to prepare the statement.

```rust

/// Polls a future once.
async fn poll_once<F: Future>(mut future: Pin<&mut F>) -> Option<F::Output> {
    poll_fn(|cx| match future.as_mut().poll(cx) {
        Poll::Ready(output) => Poll::Ready(Some(output)),
        Poll::Pending => Poll::Ready(None),
    })
    .await
}

#[tokio::test]
async fn prepare_cancellation() {
    let query = "SELECT 1 + 2";

    for poll_count_1 in 1..=5 {
        'outer: for poll_count_2 in 1..=5 {
            let pool = create_pool();
            let client = pool.get().await.unwrap();

            let mut future_1 = Box::pin(client.prepare_cached(query));
            let mut future_2 = Box::pin(client.prepare_cached(query));

            // Poll the first future a fixed number of times.
            for _ in 0..poll_count_1 {
                let opt = poll_once(future_1.as_mut()).await;
                let Some(result) = opt else {
                    continue;
                };
                result.unwrap();
                continue 'outer;
            }

            // Poll the second future a fixed number of times.
            for _ in 0..poll_count_2 {
                let opt = poll_once(future_2.as_mut()).await;
                let Some(result) = opt else {
                    continue;
                };
                result.unwrap();
                continue 'outer;
            }

            // Cancel the first future.
            drop(future_1);

            // Await the second future.
            future_2.await.unwrap();
        }
    }
}
```

divergentdave commented on postgres/src/lib.rs:
```diff
@@ -394,14 +401,59 @@ impl StatementCache {
+            query: Cow::Borrowed(query),
+            types: Cow::Borrowed(types),
+        };
+        // A cache entry is either present, in which case we can use it or
+        // absent, in which case we use a semaphore to avoid doing lots of
+        // redundant preparation work. We use an Arc<tokio::sync::Semaphore>
+        // because we can't hold the std::sync::RwLock on self.map across the
+        // await point from tokio_postgres::client::Client::prepare_typed.
```
I think we may want to drop or rephrase this second sentence. (and similarly update the commit message) There are many reasons to not just hold the RwLock; for example, it's a very coarse lock, and we'd like to be able to prepare multiple distinct statements concurrently. I think discussing alternative implementation strategies in this comment could be confusing.

Here's a suggested revision, providing an overview of the whole process.
```suggestion
        // Each map entry is empty, contains a Semaphore, or contains a prepared
        // Statement. The first call to `prepare_typed()` will insert a
        // `Semaphore`. Then, tasks will wait for a semaphore permit before
        // preparing the query, to ensure that only one task sends a `PREPARE`
        // statement at once. At both steps in the process, the map will be read
        // via a read lock first, then again via a write lock, before updating
        // the entry.
```

tgeoghegan commented:
As discussed earlier, I took a look at adding a `tokio::sync::RwLock` alongside the `std::sync::RwLock<HashMap<StatementCacheKey<'static>, Either>>`, in the hopes that this would be simpler. It is, but only if we introduce a global lock for preparing statements, which I think is a bad idea. If we introduce a lock for each `StatementCacheKey`, then we're basically back to the same thing that's in this PR.

bikeshedder committed:
Fix clickhouse pool URL and add marker to "official" deadpool-* crates

bikeshedder committed:
Remove example-postgres-hyper from CI tests

bikeshedder committed:
Do not run tests for disabled redis features

bikeshedder committed:
Update Rust version of Dev Container to 1.88

matt961 committed:
Update lapin dependency to version 3

bikeshedder committed:
Update redis dependency to version 0.32

bikeshedder committed:
Release deadpool-redis 0.22.0

bikeshedder committed:
Add *-reactor-trait dependencies

bikeshedder committed:
Release deadpool-lapin 0.13.0

bikeshedder committed:
Remove workspace and move crates to crates/*

bikeshedder committed:
Fix clippy warnings

kriswuollett committed:
Update rusqlite to 0.37.0 (#430)

bikeshedder committed:
Fix branch name in README

bikeshedder committed:
Release deadpool-sqlite 0.12.0

bikeshedder committed:
Fix docs.rs documentation compilation

This closes #432

bikeshedder committed:
Release deadpool-sqlite 0.12.1

bikeshedder committed:
Remove stray link from docs

bikeshedder committed:
Add missing re-exported features

bikeshedder committed:
Add VSCode workspace config

bikeshedder committed:
Mimick default features of diesel

bikeshedder committed:
Fix minimal version requirements

bikeshedder committed:
Update Dev Container

bikeshedder committed:
Split CI workflows by crate

bikeshedder committed:
Add missing re-export of TimeoutType

This closes #406

bikeshedder committed:
Replace "impl Default" by "#[derive(Default)]" for enums

bikeshedder committed:
Remove RecyclingMethod::Fast which does not detect broken connections

This closes #395

bikeshedder committed:
Add unique ID to objects

This closes #310

bikeshedder committed:
Remove `restart: unless-stopped` from services

bikeshedder committed:
Remove network_mode comment

bikeshedder committed:
Extract Manager, Pool and Object into separate modules

bikeshedder committed:
Add WeakPool

bikeshedder committed:
Fix example code

bikeshedder committed:
Add WeakPool to re-exports

bikeshedder committed:
Install yq via docker image into Dev Container

bikeshedder committed:
Add deadpool-libsql

bikeshedder committed:
Fix MSRV check broken by multiple versions of the same crate

This is just a workaround and should be fixed by proper
parsing of the semantic versions.

bikeshedder committed:
Remove stray comment

bikeshedder committed:
Release deadpool-libsql 0.1.0

Keruspe committed:
lapin: fix build with tokio on windows

lapin currently doesn't support tokio reactor on windows

Fixes #436

Keruspe committed:
lapin: update tokio-reactor-trait

bikeshedder committed:
Update CHANGELOG

bikeshedder committed:
Run CI checks on windows, too

bikeshedder committed:
Enable integration checks for deadpool-r2d2

bikeshedder committed:
Run workflows when the CI config is changed

bikeshedder committed:
Pin windows-2025 and fix deadpool-diesel[mysql] check

bikeshedder committed:
Release deadpool-lapin 0.13.1

bikeshedder committed:
Fix links in CHANGELOG

bikeshedder committed:
Fix deadpool-tiberius entry

Commit:
feat(deadpool): cache num_cpus for better init performance

bikeshedder committed:
Fix default pool max size

bikeshedder committed:
Improve documentation of Object::id

bikeshedder committed:
Add wrapper for ObjectId making it an opaque data type

bikeshedder committed:
Release deadpool 0.12.3

bikeshedder committed:
Fix deadpool-memcached

bikeshedder committed:
Release deadpool-memcached 0.4.0

bikeshedder committed:
Update github actions

bikeshedder committed:
Reexport hickory-dns

bikeshedder committed:
Fix repository URLs

bikeshedder committed:
Add build status badge to README

bikeshedder committed:
Add VSCode workspace file

bikeshedder committed:
Update rust version to 1.90 in dev container

werekraken committed:
Fix `post_create()` prefix in managed module documentation

werekraken committed:
Fix `RecycleError::message()` summary in documentation

tgeoghegan committed:
review comments

tgeoghegan committed:
cargo fmt
